### PR TITLE
Add several macros for making the runtime easier to maintain

### PR DESCRIPTION
**Pull request by TyOverby**

### CODE REVIEW

**Comment by TyOverby** [comment]
@hhugo I plan on expanding this PR to include actual runtime modifications.

File: compiler/lib/driver.ml [hunk 1]
```diff
@@ -286,21 +286,29 @@ let link ~standalone ~linkall ~export_runtime (js : Javascript.source_elements)
 
 let macro recurse fallthrough =
   let module J = Javascript in
+  let zero, one = J.ENum "0", J.ENum "1" in
   function
+  | "BLOCK", [_] ->
+    failwith (Format.sprintf "macro BLOCK called with inappropriate arguments")
   | "BLOCK", tag :: args ->
```
Comment (hhugo) [inline compiler/lib/driver.ml]: You can use
`"BLOCK", tag :: ( _ :: _ as args)`
and get rid of the match on "BLOCK" above.

File: compiler/lib/driver.ml [hunk 2]
```diff
@@ -286,21 +286,29 @@ let link ~standalone ~linkall ~export_runtime (js : Javascript.source_elements)
...
       let adjusted = J.ENum (string_of_int (idx + 1)) in
       J.EAccess (recurse e, adjusted)
   | "FIELD", [e; idx] ->
-      let adjusted = J.EBin (J.Plus, J.ENum "1", recurse idx) in
+      let adjusted = J.EBin (J.Plus, one, recurse idx) in
       J.EAccess (recurse e, adjusted)
-  | "ISBLOCK", [_] -> failwith "what is this?"
+  | "ISBLOCK", [e] -> J.EBin
+            ( J.EqEq
+            , J.EUn (J.Typeof, J.EAccess (recurse e, zero))
```
Comment (hhugo) [inline compiler/lib/driver.ml]: `ISBLOCK(x) => typeof x[0] == "number"` ?
I would expect 
`ISBLOCK(x) => typeof x != "number"`

File: compiler/lib/macro.ml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+   less efficient code is generated. *)
+
+let is_block e = J.EBin (J.NotEqEq, J.EUn (J.Typeof, e), J.EStr ("number", `Bytes))
+
+let is_immediate e = J.EBin (J.EqEqEq, J.EUn (J.Typeof, e), J.EStr ("number", `Bytes))
+
+let field e i =
+  match i with
+  | J.ENum n ->
+      let idx = int_of_string n in
```
Comment (hhugo) [inline compiler/lib/macro.ml]: We should validate the string (will be incorrect with the JavaScript octal notation)

File: compiler/lib/macro.ml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    inherit Js_traverse.map as super
+
+    method expression x =
+      let module J = Javascript in
+      match x with
+      | J.ECall (J.EVar (J.S {name; _}), args, _) -> (
+        match name, args with
+        | "BLOCK", J.ENum tag :: (_ :: _ as args) ->
+            make_block
+              ~tag:(int_of_string tag)
```
Comment (hhugo) [inline compiler/lib/macro.ml]: We should validate the string (will be incorrect with the JavaScript octal notation)

File: compiler/lib/macro.mli [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+val tag : Javascript.expression -> Javascript.expression
+
+val is_block : Javascript.expression -> Javascript.expression
+
+val is_immediate : Javascript.expression -> Javascript.expression
+
+val length : Javascript.expression -> Javascript.expression
+
+val field : Javascript.expression -> Javascript.expression -> Javascript.expression
```
Comment (hhugo) [inline compiler/lib/macro.mli]: We should probably add helpers for array literal, array access and update

**Comment by hhugo** [comment]
@TyOverby, any idea for a better name for macro.ml ?

**Comment by hhugo** [comment]
@TyOverby, I've made the following changes:
- abstracted JavaScript numbers (previously exposed as string)
- handle the case where javascript numbers use the octal notation. 
- renamed macro.ml -> mlvalue.ml
- Used mlvalue to generate code in generate.ml

I'm done with changes and code review.

**Comment by hhugo** [comment]
I've extracted some of my changes and rebased your PR on top.

**Comment by TyOverby** [comment]
Looks great; thanks @hhugo

**Pull request merged**